### ISSUE TITLE
Clarification on Family.debugKeys (also request to change it to Family.keys)

### ISSUE THREAD

**Issue by daadu** [rrousselGit/river_pod#192]
**Describe what scenario you think is uncovered by the existing examples/articles**
`Family.debugKeys` is described vaguely as "debug-only" in the [docs](https://pub.dev/documentation/riverpod/latest/all/Family/debugKeys.html). 

# Possible Solution
- Clarify why they are "debug-only" and scenarios describing why they should not be used.
- I dig in the code, it seems that it is ok to use the `debugKeys` list, why cant we make remove the prefix "debug" (obviously in backward compatible way with proper deprecation warning): `List debugKeys` -> `Set keys` (Notice the use of `Set` instead of `List` since are easier to lookup), this could be really handy to check what `keys` are "active" (in case of `.autoDispose`) in the `family`.

**Comment by daadu** [comment]
Also I am new to Riverpod, I might be missing something. Feel free to correct me.

**Comment by rrousselGit** [comment]
This list is unreliable and probably be removed entirely

Could you describe what is your use case for it?

**Comment by daadu** [comment]
I have two `AutoDisposeStateNotifierProviderFamily ` provider:
  - `_getProvider` : family that returns "single item" of a model, based on id
  - _getManyProvider`: family that return "multiple item" of model, based on a query object

Both of them are wrapped around `ModelProvider` which itself is provided by simple `Provider`.

Now the use case is, when any item from `_getProvider` is update/delete i need to rebuild/refresh/notify all active in `_getManyProvider` family.

This is a very common scenario I guess, for eg in `TodoApp` with `TodoListPage` and `TodoDetailPage` (with edit, delete options), any changes in `TodoDetailPage` should be reflected on `TodoListPage` appropriately.

**Comment by daadu** [comment]
Hope you get it. 

I am trying to build a pub package around it, that could be used to create the "data layer" of the app.

**Comment by daadu** [comment]
The class that I have worked on so far, check the `_refreshItem` and `_refreshList` methods for the use case

```dart
class ModelProvider<T extends BaseModel, Q, E> {
  final ProviderReference ref;
  final BaseModelRepo<T, Q, E> repo;

  final AutoDisposeStateNotifierProviderFamily<_ListNotifier<T, Q>, Q>
      _getManyProvider;
  final AutoDisposeStateNotifierProviderFamily<_ItemNotifier<T>, String>
      _getProvider;

  ModelProvider(this.ref, this.repo)
      : _getManyProvider =
            StateNotifierProvider.autoDispose.family((ref, query) {
          return _ListNotifier<T, Q>(repo, query);
        }),
        _getProvider = StateNotifierProvider.autoDispose
            .family<_ItemNotifier<T>, String>(
                (ref, id) => _ItemNotifier(repo, id));

  AutoDisposeStateNotifierProviderFamily<_ListNotifier<T, Q>, Q> get getMany =>
      _getManyProvider;

  AutoDisposeStateNotifierProviderFamily<_ItemNotifier<T>, String> get get =>
      _getProvider;

  void _refreshItem(T item) {
    // refresh the ItemNotifier, if exists
    if (get.debugKeys.contains(item.id)) {
      ref.read(get(item.id)).refresh();
    }
  }

  void _refreshList([T item]) {
    // refresh all ListNotifier, where item is present
    // TODO refresh the ListNotifier in which the item is present
    for (final f in getMany.debugKeys) {
      ref.read(getMany(f)).refresh();
    }
  }


  void add(T item) {
    repo.add(item);
    _refreshList();
  }

  void update(T item) {
    repo.update(item);
    _refreshItem(item);
    _refreshList(item);
  }

  void delete(T item) {
    repo.delete(item);
    _refreshList(item);
  }
}

class _ListNotifier<T extends BaseModel, Q>
    extends StateNotifier<BuiltList<T>> {
  _ListNotifier(this.repo, [this.query]) : super(repo.getMany(query));

  final BaseModelRepo<T, Q, dynamic> repo;
  final Q query;

  void refresh() {
    state = repo.getMany(query);
  }
}

class _ItemNotifier<T extends BaseModel> extends StateNotifier<T> {
  _ItemNotifier(this.repo, this.id) : super(repo.get(id));

  final BaseModelRepo<T, dynamic, dynamic> repo;
  final String id;

  void refresh() {
    state = repo.get(id);
  }
}

```

**Comment by rrousselGit** [comment]
Would the ability to dispose all the providers of a family at once work for you?

**Comment by daadu** [comment]
It would work, but I am thinking of adding logic, to only dispose those that are required - for performance sake (refetching queries from db or network)

**Comment by rrousselGit** [comment]
Could you expand on what that logic is?

**Comment by daadu** [comment]
I am not sure, but I have thought of having a `Map` (or combination of other collectable) with (`modelId` => `List<Q query>`). in the `repo` that delivers the List based on query, i write a wrapper method, that populates this map. 

Now if `Family.keys` is available it would simply the job of keeping track of active providers or else I have to maintain that manually - having a `Set(Q query)`, extending `AutoDisposeStateNotifierProviderFamily` that populates this set when `createState` is called and `onDispose` removing it from it.

**Comment by daadu** [comment]
TLDR: `Family.keys` would really help in implementing this.

Also note, if `Family.keys` is exposed then it should be made immutable (with something like `built_collection`).

**Comment by daadu** [comment]
I am aware that this "optimization" could be slower than just disposing all (and recreating), hence this could be configured by the user.

**Comment by rrousselGit** [comment]
What prevents you from manually creating this map yourself?
Like:

```dart
Set<RootProvder> _providers = {};

final provider = Provider.autoDispose((ref) {
  _providers.add(this);
  ref.onDispose(() => _providers.remove(this));
});

**Comment by daadu** [comment]
Since the providers are inside class and final, I need to initialize them:

```dart
class class ModelProvider<T extends BaseModel, Q, E>{
  ...
  final AutoDisposeStateNotifierProviderFamily<_ListNotifier<T, Q>, Q>
      _getManyProvider;
  final AutoDisposeStateNotifierProviderFamily<_ItemNotifier<T>, String>
      _getProvider;

 ModelProvider(this.ref, this.repo)
      : _getManyProvider =
            StateNotifierProvider.autoDispose.family((ref, query) {
          return _ListNotifier<T, Q>(repo, query);
        }),
        _getProvider = StateNotifierProvider.autoDispose
            .family<_ItemNotifier<T>, String>((ref, id) {
          return _ItemNotifier(repo, id);
        });
  ...
}
```

[Dart does not allow using instance member in initialiser.](https://dart.dev/tools/diagnostic-messages#implicit_this_reference_in_initializer) Therefore something like following is not possible:

```dart
class ModelProvider<T extends BaseModel, Q> {
  ... 
 final Set<String> _activeItemProviderKeys = {};

 ModelProvider(this.ref, this.repo)
      : _getManyProvider = ...
        _getProvider = StateNotifierProvider.autoDispose
            .family<_ItemNotifier<T>, String>((ref, id) {
          _activeItemProviderKeys.add(id);
          ref.onDispose(() => _activeItemProviderKeys.remove(id));
          return _ItemNotifier(repo, id);
        });
  ...
}
```

Possible workaround could be not making `_getProvider` and `_getManyProvider` `final`, is that OK? since `ModelProvider` itself will be provided with `Provider`. I did not "feel" right with using non-final attributes inside a class that will be "provided", globally. But thinking again I think this could be safe since I am just exposing the private member via just "getter" to outer world. What do you think @rrousselGit? Do I have you blessings?

**Comment by rrousselGit** [comment]
That's fine, as long as you don't change the value

**Comment by daadu** [comment]
Closing this issue, since populating `keys` of the family provider can be done manually.
